src/analyses/local_safe_pointers.cpp

```diff
@@ -0,0 +1,239 @@
+/*******************************************************************\
+
+Module: Local safe pointer analysis
+
+Author: Diffblue Ltd
+
+\*******************************************************************/
+
+/// \file
+/// Local safe pointer analysis
+
+#include "local_safe_pointers.h"
+
+#include <langapi/language_util.h>
+#include <util/expr_iterator.h>
+
+/// If `expr` is of the form `x != nullptr`, return x. Otherwise return nil.
+static exprt get_null_checked_expr(const exprt &expr)
+{
+  if(expr.id() == ID_notequal)
+  {
+    exprt op0 = expr.op0(), op1 = expr.op1();
+    if(op0.type().id() == ID_pointer &&
+       op0 == null_pointer_exprt(to_pointer_type(op0.type())))
+    {
+      std::swap(op0, op1);
+    }
+
+    if(op1.type().id() == ID_pointer &&
+       op1 == null_pointer_exprt(to_pointer_type(op1.type())))
+    {
+      while(op0.id() == ID_typecast)
+        op0 = op0.op0();
+      return op0;
+    }
+  }
+
+  return nil_exprt();
```

Comment:
To comply with my recommendation of returning a `const exprt &` do `return static_cast<const exprt&>(get_nil_irep());`

Reply:
I've just used a pointer as a "poor man's optional," but otherwise done as you suggest.